File: react-native-pytorch-core/ios/JavaScript/JSContext.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

@objc(PTLJSContext)
public class JSContext: NSObject {

    enum JSContextError: Error {
        case invalidParam
    }

    public static let idKey = "ID"

    public static var refs: [String: NativeJSRef] = [:]

    public static func setRef(ref: NativeJSRef) -> String {
        let refId = UUID().uuidString
        JSContext.refs[refId] = ref
        return refId
    }

    public static func getRef(refId: String) throws -> NativeJSRef {
        guard let unwrappedNativeJSRef = JSContext.refs[refId] else { throw JSContextError.invalidParam }
        return unwrappedNativeJSRef
    }

    public static func get(jsRef: [ String: String ]) throws -> NativeJSRef {
        guard let refId = jsRef[idKey] else { throw JSContextError.invalidParam }
        return try JSContext.getRef(refId: refId)
    }

    @objc
    public static func release(jsRef: [ String: String ]) throws {
        guard let refId = jsRef[idKey] else { throw JSContextError.invalidParam }
        let removedJSRef = refs.removeValue(forKey: refId)
        try removedJSRef?.release()
    }

    public static func wrapObject(object: Any) -> NativeJSRef {
        return NativeJSRef(object: object)
    }

    public static func unwrapObject(jsRef: [ String: String ]) throws -> Any {
        guard let refId = jsRef[idKey] else { throw JSContextError.invalidParam }
        let ref = try JSContext.getRef(refId: refId)
        return ref.getObject()
    }

    public class NativeJSRef {
        // initialized mId and mJSRef to empty values to allow self to be used to set id in init()
        private var mId: String? = ""
        private var mObject: Any?
        private var mJSRef: [String: String]? = [:]

        init(object: Any) {
            mObject = object
            mId = JSContext.setRef(ref: self)
            mJSRef?[JSContext.idKey] = mId
        }

        public func getJSRef() -> [String: String] {
            return mJSRef ?? [:]
        }

        public func getObject() -> Any {
            return mObject as Any
        }

        public func release() throws {
            if let image = mObject as? Image {
                try image.close()
            }
            mId = nil
            mObject = nil
            mJSRef = nil
        }
    }
}

File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/javascript/JSContext.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.javascript;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import java.util.UUID;
import java.util.WeakHashMap;

public class JSContext {

  protected static final String ID_KEY = "ID";

  private static final WeakHashMap<String, NativeJSRef> refs = new WeakHashMap<>();

  public static String setRef(NativeJSRef ref) {
    String id = UUID.randomUUID().toString();
    JSContext.refs.put(id, ref);
    return id;
  }

  public static NativeJSRef getRef(String id) {
    return JSContext.refs.get(id);
  }

  public static NativeJSRef get(ReadableMap jsRef) {
    String id = jsRef.getString(ID_KEY);
    return JSContext.getRef(id);
  }

  public static void release(String id) throws Exception {
    NativeJSRef ref = JSContext.getRef(id);
    ref.release();
    JSContext.refs.remove(id);
  }

  public static void release(ReadableMap jsRef) throws Exception {
    String id = jsRef.getString(ID_KEY);
    JSContext.release(id);
  }

  public static NativeJSRef wrapObject(Object object) {
    return new NativeJSRef(object);
  }

  public static <T> T unwrapObject(ReadableMap jsRef) {
    String id = jsRef.getString(ID_KEY);
    NativeJSRef ref = JSContext.getRef(id);
    return (T) ref.getObject();
  }

  public static final class NativeJSRef {

    private String mId;
    private Object mObject;
    private WritableMap mJSRef;

    protected NativeJSRef(Object object) {
      mObject = object;
      mId = JSContext.setRef(this);
      mJSRef = Arguments.createMap();
      mJSRef.putString(JSContext.ID_KEY, mId);
    }

    public WritableMap getJSRef() {
      return mJSRef;
    }

    public Object getObject() {
      return mObject;
    }

    /**
     * This method is deliberately private to only allow the outer class to release a NativeJSRef
     * instance.
     *
     * @throws Exception Forwarding any exception that is raised during the close call of the
     *     wrapped object.
     */
    private void release() throws Exception {
      if (mObject instanceof AutoCloseable) {
        ((AutoCloseable) mObject).close();
      }
      // Remove reference to id, which will release the associated NativeJSRef object stored as
      // value for the ID key.
      mId = null;
      mJSRef = null;
      mObject = null;
    }
  }
}
